src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideStreamMemoryBasedAsyncOverrides.cs

```diff
@@ -0,0 +1,208 @@
+            var cancellationTokenType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingCancellationToken);
+
+            if (memoryOfByteType is null || readOnlyMemoryOfByteType is null || streamType is null || cancellationTokenType is null)
+            {
+                requiredSymbols = default;
+                return false;
+            }
+
+            var readAsyncArrayMethod = GetOverload(streamType, ReadAsyncName, byteArrayType, int32Type, int32Type, cancellationTokenType);
+            var readAsyncMemoryMethod = GetOverload(streamType, ReadAsyncName, memoryOfByteType, cancellationTokenType);
+            var writeAsyncArrayMethod = GetOverload(streamType, WriteAsyncName, byteArrayType, int32Type, int32Type, cancellationTokenType);
+            var writeAsyncMemoryMethod = GetOverload(streamType, WriteAsyncName, readOnlyMemoryOfByteType, cancellationTokenType);
+
+            if (readAsyncArrayMethod is null || readAsyncMemoryMethod is null || writeAsyncArrayMethod is null || writeAsyncMemoryMethod is null)
+            {
+                requiredSymbols = default;
+                return false;
+            }
+
+            requiredSymbols = new RequiredSymbols(
+                streamType, memoryOfByteType, readOnlyMemoryOfByteType,
+                readAsyncArrayMethod, readAsyncMemoryMethod,
+                writeAsyncArrayMethod, writeAsyncMemoryMethod);
+            return true;
+        }
+
+        private static IMethodSymbol? GetOverload(ITypeSymbol containingType, string methodName, params ITypeSymbol[] argumentTypes)
+        {
+            return containingType.GetMembers(methodName)
+                .SingleOrDefault(symbol => symbol is IMethodSymbol m && IsMatch(m, argumentTypes)) as IMethodSymbol;
+
+            static bool IsMatch(IMethodSymbol method, ITypeSymbol[] argumentTypes)
+            {
+                if (method.Parameters.Length != argumentTypes.Length)
+                    return false;
+
+                for (int index = 0; index < argumentTypes.Length; ++index)
+                {
+                    if (!argumentTypes[index].Equals(method.Parameters[index].Type, SymbolEqualityComparer.Default))
+                        return false;
+                }
+
+                return true;
+            }
+        }
+
+        //  If the specified type overrides the specified method on its immediate base class, returns the method symbol representing the 
+        //  overriding method. Returns null if the specified type does not override the specified method.
+        private static IMethodSymbol? GetSymbolForOverridingMethod(ITypeSymbol derivedType, IMethodSymbol overriddenMethod)
+        {
+            RoslynDebug.Assert(derivedType.BaseType.Equals(overriddenMethod.ContainingType, SymbolEqualityComparer.Default));
+
+            return derivedType.GetMembers(overriddenMethod.Name)
+                .SingleOrDefault(x =>
```

Comment:
Can it happen there are multiple symbols with these criteria, hence the call to `SingleOrDefault` throws an exception?

Reply:
It should never happen. We check if

1. The method has a specific name,
2. It is an overriding method (which means its original definition must be a virtual method, which rules out any method-hiding)
3. The types of its arguments match exactly.

I don't see any way in C# or VB for a class to have two methods that match all of those criteria. I used `SingleOrDefault` deliberately to document the impossibility of the above scenario.

Reply:
@NewellClark Yes, this most likely shouldn't happen for valid code. But my concern was about "invalid" code. The analyzers are run against code that has compile error.

Consider checking what would happen if the same method is just repeated (which will have a compile-error). Will GetMembers return repeated methods for this error scenario?

Reply:
@Youssef1313 Good catch. Just tested it and it does indeed crash. Working on a fix right now.